nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -254,6 +307,27 @@ def training_step(self, batch, batch_idx):
         decoder_seq_length = batch_for_pipeline[1].size(1)
         tensor_shape = [encoder_seq_length, get_micro_batch_size(), self.cfg.encoder.hidden_size]
 
+        # handle asynchronous grad reduction
+        if self.with_distributed_adam:
+            if self.megatron_amp_o2:
+                # copy grads to main grad
+                custom_sync_context_handler = lambda: self._optimizer.no_sync(greedy_grad_copy=True)
+            else:
+                # keep grad tensors around
+                custom_sync_context_handler = lambda: self._optimizer.no_sync(greedy_grad_copy=False)
+        else:
+            if (
+                self.megatron_amp_o2
+                and self.cfg.get('pipeline_model_parallel_size', 1) == 1
+                and not self.cfg.get('sequence_parallel', False)
+            ):
+                custom_sync_context_handler = self._optimizer.no_sync
+            else:
+                # TODO: enable async grad all reduce with O1/autocast
+                # mixed precision training, with pipeline parallelism,
+                # or with sequence parallelism
+                custom_sync_context_handler = None
+
```

Comment:
Can we move this to the base model? GPT has the same code.

Reply:
In the `main` branch, the non-distopt behavior with [GPT](https://github.com/NVIDIA/NeMo/blob/86565741f22d0c19da07680e2f40bc207c613206/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L241) and [T5](https://github.com/NVIDIA/NeMo/blob/86565741f22d0c19da07680e2f40bc207c613206/nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py#L257) is subtly different. In particular, T5 doesn't pass in a context if pipeline parallelism is enabled. We could make the behavior between GPT and T5 consistent, but I think that would be better in a standalone PR so it's easier to identify any regressions.